mono/sgen/sgen-gc.c

```diff
@@ -2113,7 +2093,7 @@ major_start_concurrent_collection (const char *reason)
 	// FIXME: store reason and pass it when finishing
 	major_start_collection (TRUE, NULL);
 
-	gray_queue_redirect (&gray_queue);
+	sgen_workers_take_from_queue_and_awake (&gray_queue);
 
```

Comment:
When we enable redirection for the gc_thread_gray_queue we move all the sections to the workers distribute gray queue and we set a callback that moves any newly allocated section there also. And since only the gc thread enqueues to it's own gray queue, I'm assuming it should stay empty.